### PR TITLE
All: Do not auto-enable deprecated packages

### PR DESCRIPTION
**Pull request by sebrowne**
@trilinos/framework 

Set all packages that will be deprecated next year to `ST` (secondary tested).

Also default `Trilinos_ENABLE_SECONDARY_TESTED_CODE` to `OFF`.

Now, TriBITS will not automatically enable any of the to-be-deprecated packages if `Trilinos_ENABLE_SECONDARY_TESTED_CODE` is `OFF`.

Most of the PR builds enable secondary tested code, so this should be fairly safe.

Summary: Do NOT automatically enable deprecated packages, even if `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES` is `ON` (default).

### This is a change for user-facing behavior.  If a user was depending on automatic package enablement for explicit behavior, they will need to specify that from now on (e.g., they were enabling Thyra and needed ThyraEpetraAdapters, they will now need to explicitly enable ThyraEpetraAdapters).

Tagging members of the operational leadership team, Curt (Product Manager), along with Ross (in case I missed anything about how ST works).

## Motivation
Now that a warning message is printed if deprecated packages are enabled (#12746), we want to not auto-enable any of those packages (somewhat contradictory behavior).

In addition, in preparation for behavior as in #12828 , we don't want to present users with quite as much in-your-face warning behavior if they did not mean to enable deprecated packages (combination of `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES` defaulting to `ON` and `Trilinos_ENABLE_SECONDARY_TESTED_CODE` defaulting to `ON`.

## Testing
I tested the following configurations with the described behavior:

`cmake -DTrilinos_ENABLE_Thyra=ON ..`
This enabled Thyra, but NOT Epetra or ThyraEpetraAdapters.

```
-- NOTE: Not Setting Trilinos_ENABLE_ThyraEpetraAdapters=ON even though Thyra has an optional dependence on ThyraEpetraAdapters because Trilinos_ENABLE_SECONDARY_TESTED_CODE=OFF
-- NOTE: Not Setting Trilinos_ENABLE_Epetra=ON even though TpetraCore has an optional dependence on Epetra because Trilinos_ENABLE_SECONDARY_TESTED_CODE=OFF
```

`cmake -DTrilinos_ENABLE_Thyra=ON -DTrilinos_ENABLE_SECONDARY_TESTED_CODE=ON ..`
This enabled Thyra, Epetra, and ThyraEpetraAdapters

```
-- Setting subpackage enable Trilinos_ENABLE_ThyraEpetraAdapters=ON because parent package Trilinos_ENABLE_Thyra=ON
-- Setting Trilinos_ENABLE_Epetra=ON because ThyraEpetraExtAdapters has a required dependence on Epetra
```

`cmake -DTrilinos_ENABLE_Thyra=ON -DTrilinos_ENABLE_Epetra=ON ..`
This enabled Thyra and Epetra, but NOT ThyraEpetraAdapters.

```
-- NOTE: Not Setting Trilinos_ENABLE_ThyraEpetraAdapters=ON even though Thyra has an optional dependence on ThyraEpetraAdapters because Trilinos_ENABLE_SECONDARY_TESTED_CODE=OFF
```

`cmake -DTrilinos_ENABLE_Thyra=ON -DTrilinos_ENABLE_Epetra=ON -DTrilinos_ENABLE_SECONDARY_TESTED_CODE=ON ..`
This enabled Thyra, Epetra, and ThyraEpetraAdapters.

```
-- Setting subpackage enable Trilinos_ENABLE_ThyraEpetraAdapters=ON because parent package Trilinos_ENABLE_Thyra=ON
```

### CODE REVIEW

**Review by mperego — APPROVED** [review]
Looks good to me. Thanks @sebrowne !

**Comment by mperego** [comment]
> `cmake -DTrilinos_ENABLE_Thyra=ON -DTrilinos_ENABLE_Epetra=ON ..`
> This enabled Thyra and Epetra, but NOT ThyraEpetraAdapters.

This might create mild issues to applications, but it's a good way to further signal that they are using packages that will be eventually removed.

File: ProjectName.cmake [hunk 1]
```diff
@@ -1,6 +1,6 @@
 SET(PROJECT_NAME Trilinos)
 
-SET(${PROJECT_NAME}_ENABLE_SECONDARY_TESTED_CODE_DEFAULT  ON)
+SET(${PROJECT_NAME}_ENABLE_SECONDARY_TESTED_CODE_DEFAULT  OFF)
```
Comment (bartlettroscoe) [inline ProjectName.cmake:3]: This is a break in backward compatibility.  This will break many users configure scripts.

**Review by bartlettroscoe — COMMENTED** [review]
> Summary: Do NOT automatically enable deprecated packages, even if `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES` is `ON` (default).

I don't see how the changes in this PR has that effect in all cases.   For example, if someone changes an ST package then the PR build that has `Trilinos_ENABLE_SECONDARY_TESTED_CODE=OFF` will still enable those changed ST packages.  (It just will not enable the intra-package dependencies for those enabled and changed ST packages.)  If you want to have a PR build that disables all deprecated packages, I think you need to hard disable them in the configuration settings for that build.

> > `cmake -DTrilinos_ENABLE_Thyra=ON -DTrilinos_ENABLE_Epetra=ON ..`
> > This enabled Thyra and Epetra, but NOT ThyraEpetraAdapters.
> 
> This might create mild issues to applications, but it's a good way to further signal that they are using packages that will be eventually removed. 

That is going to be confusing to some users and this change definingly breaks backward compatibility for everyone's builds that involve these deprecated packages.

I think it might be worth taking a step back and considering what the goals are here related to these deprecated packages and making the transition for users and developers to be smoother.

**Comment by sebrowne** [comment]
> > Summary: Do NOT automatically enable deprecated packages, even if `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES` is `ON` (default).
> 
> I don't see how the changes in this PR has that effect in all cases. For example, if someone changes an ST package then the PR build that has `Trilinos_ENABLE_SECONDARY_TESTED_CODE=OFF` will still enable those changed ST packages. (It just will not enable the intra-package dependencies for those enabled and changed ST packages.) If you want to have a PR build that disables all deprecated packages, I think you need to hard disable them in the configuration settings for that build.

That's a PR-specific behavior, since we explicitly tell TriBITS to turn on packages that we have identified as being changed (but yes, for PRs that touch ST packages, I think the behavior stays the same).  For _users_, the behavior does change in that if they were explicitly enabling superpackages and expecting all subpackages to be enabled.  I think it's low impact honestly, and it would have been better practice for them to explicitly enable the desired subpackage in the first place.

File: ProjectName.cmake (hunk 1 above)
Comment (sebrowne) [reply ProjectName.cmake:3]: Agreed.  This is the most dangerous part of this change (and is what the large text in the description is intended to convey).

**Comment by bartlettroscoe** [comment]
> For _users_, the behavior does change in that if they were explicitly enabling superpackages and expecting all subpackages to be enabled. I think it's low impact honestly, and it would have been better practice for them to explicitly enable the desired subpackage in the first place.

Users that want the old behavior can just set `Trilinos_ENABLE_SECONDARY_TESTED_CODE=OFF` in their configure scripts.

But given this is a break in backward compatibility, should there not be an entry added to the RELEASE_NOTES file?

**Comment by mperego** [comment]
@sebrowne, what's holding this up?

**Comment by jwillenbring** [comment]
> @sebrowne, what's holding this up?

I threw a retest label on so that isn't an issue at least.

**Comment by achauphan** [comment]
> Status Flag 'Pull Request AutoTester' - Failure: Job Trilinos_PR_gcc-8.3.0 on Jenkins server

Looking into this

**Comment by sebrowne** [comment]
> But given this is a break in backward compatibility, should there not be an entry added to the RELEASE_NOTES file?

@bartlettroscoe done

**Comment by sebrowne** [comment]
After discussion within the leadership team and at the developer's meeting, this was decided to be an acceptable approach.  I will craft an email as well notifying users of the behavior change.

**Review by mperego — APPROVED** [review]

**Pull request merged**